lib/sqlalchemy/dialects/postgresql/base.py

```diff
@@ -3643,20 +3643,152 @@ def get_multi_columns(
-        attype_pattern = re.compile(r"\(.*\)")
-        charlen_pattern = re.compile(r"\(([\d,]+)\)")
-        args_pattern = re.compile(r"\((.*)\)")
-        args_split_pattern = re.compile(r"\s*,\s*")
-
-        def _handle_array_type(attype):
-            return (
-                # strip '[]' from integer[], etc.
-                array_type_pattern.sub("", attype),
-                attype.endswith("[]"),
+    _format_type_args_pattern = re.compile(r"\((.*)\)")
+    _format_array_spec_pattern = re.compile(r"((?:\[\])*)$")
+
+    def _reflect_type(
+        self,
+        format_type: Optional[str],
+        domains: dict[str, ReflectedDomain],
+        enums: dict[str, ReflectedEnum],
+        type_description: Optional[str] = None,
+    ) -> sqltypes.TypeEngine[Any]:
+        """
+        Attempts to reconstruct a column type defined in ischema_names based
+        on the information available in the format_type.
+
+        If the `format_type` cannot be associated with a known `ischema_names`,
+        then it is treated it is treated a reference to a known PostgreSQL
+        named `ENUM` or `DOMAIN` type.
+        """
+        type_description = type_description or "unknown type"
+        if format_type is None:
+            util.warn(
+                "PostgreSQL format_type() returned NULL for %s"
+                % type_description
             )
+            return sqltypes.NULLTYPE
+
+        attype_args_match = self._format_type_args_pattern.search(format_type)
+        if attype_args_match and attype_args_match.group(1):
+            attype_args = re.split(r"\s*,\s*", attype_args_match.group(1))
+        else:
+            attype_args = ()
+
+        match_array_dim = self._format_array_spec_pattern.search(format_type)
+        # Each "[]" in array specs corresponds to an array dimension
+        array_dim = len(match_array_dim.group(1) or "") // 2
+
+        # Remove all parameters and array specs from format_type to obtain an
+        # ischema_name candidate
+        attype = self._format_type_args_pattern.sub("", format_type)
+        attype = self._format_array_spec_pattern.sub("", attype)
+
+        schema_type = self.ischema_names.get(attype.lower(), None)
+        args, kwargs = (), {}
+
+        if attype == "numeric":
+            if len(attype_args) >= 2:
+                precision, scale = map(int, attype_args)
+                args = (precision, scale)
+
+        elif attype == "double precision":
+            args = (53,)
+
+        elif attype == "integer":
+            args = ()
+
+        elif attype in ("timestamp with time zone", "time with time zone"):
+            kwargs["timezone"] = True
+            if len(attype_args) >= 1:
+                kwargs["precision"] = int(attype_args[0])
 
+        elif attype in (
+            "timestamp without time zone",
+            "time without time zone",
+            "time",
+        ):
+            kwargs["timezone"] = False
+            if len(attype_args) >= 1:
```

Comment:
**Federico Caselli** (CaselIT) wrote:

==

View this in Gerrit at https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5030